manual/src/cmds/tsan.etex

```diff
@@ -0,0 +1,227 @@
+unrelated synchronizing operations.
+
+\section{s:tsan-performance}{Performance implications}
+
+TSan instrumentation imposes a non-negligible cost at runtime. Empirically,
+this cost has been observed to cause a slowdown which can range from 2x to 7x.
+One of the main factors of high slowdowns is frequent of access to mutable
+data (in contrast, reads from immutable data are not instrumented). The memory
+consumption is also increased, by a factor between 4 and 7. TSan also allocates
+very large amounts of virtual memory (but uses only a fraction of it).
+
+\section{s:tsan-false-neg-false-pos}{False negatives and false positives}
+
+As illustrated by the previous example, TSan will only report the data races
+effectively encountered during execution. Another important caveat is that TSan
+remembers only a finite number of memory accesses per memory location and per
+thread. At the time of writing, this number is 4. Data races involving a
+forgotten access will not be detected. Lastly, the
+\href{https://github.com/google/sanitizers/wiki/ThreadSanitizerAlgorithm}{documentation
+of TSan} states that there is a tiny probability to miss a race if two threads
+access the same location at the same time. These are the only three cases that
+can cause TSan to miss data races.
+
+For data races between two memory accesses made from OCaml code, TSan does not
+produce false positives; that is, TSan will not emit spurious reports.
+
+When mixing OCaml and C code, through the use of C primitives, the very notion
+of false positive becomes less clear, as it involves two memory models.
+However, TSan should behave mostly as one would expect: non-atomic reads and writes
+in C will race with non-atomic reads and writes in OCaml, and C atomics will
+not race with OCaml atomics. There is one theoretical possibility of false
+positive: if a \texttt{value} is initialized from C without using
+\texttt{caml_initialize} (which is allowed under the condition that the GC does
+not run between the allocation and the write, see Chapter~\ref{c:intf-c}) and a
+conflicting access is made later by another thread. This does not constitute a
+data race, but TSan may report it as such.
+
+\section{s:tsan-runtime-flags}{Runtime options}
+
+TSan supports a number of configuration options at runtime using the
+\texttt{TSAN\_OPTIONS} environment variable. Notably, it allows to suppress
+some data races from TSan reports. See the
+\href{https://github.com/google/sanitizers/wiki/ThreadSanitizerFlags}{documentation
+of TSan flags} and the
+\href{https://github.com/google/sanitizers/wiki/SanitizerCommonFlags}{documentation
+of flags common to all sanitizers} for more information. Suppressing data race
+reports is useful for intentional races or libraries that cannot be fixed.
+
+For example, to suppress reports originating from functions in the OCaml module
+\texttt{My_module}, one could set
+\texttt{TSAN\_OPTIONS} to \texttt{"suppressions=suppr.txt"}, with
```

Comment:
Can you write out the suppressions command explicitly like you do for `history_size` later?

Would it be

```
TSAN_OPTIONS="suppressions=suppr.txt" ./my_instrumented_program
```

with `suppr.txt` containing

```
race_top:^camlMy_module
```

Is there an external link to what is the format of `suppr.txt`?

Reply:
I wrote out the whole command example.

> Is there an external link to what is the format of suppr.txt?

It can be found in the TSan documentation page that I link in the paragraph that precedes.